envtest/ibmblockcsi_controller_test.go

```diff
@@ -0,0 +1,151 @@
+				err := k8sClient.Create(context.Background(), ibc)
+				Expect(err).NotTo(HaveOccurred())
+
+				found := &csiv1.IBMBlockCSI{}
+				key := types.NamespacedName{
+				  Name:      ibcName,
+				  Namespace: namespace,
+				}
+
+				By("Getting IBMBlockCSI object after creation")
+				Eventually(func() (*csiv1.IBMBlockCSI, error) {
+				  err := k8sClient.Get(context.Background(), key, found)
+				  return found, err
+				}, timeout, interval).ShouldNot(BeNil())
+
+				By("Getting CSIDriver")
+				cd := &storagev1.CSIDriver{}
+				Eventually(func() (*storagev1.CSIDriver, error) {
+				  err := k8sClient.Get(context.Background(),
+				  	testsutil.GetResourceKey(config.DriverName, "", ""), cd)
+				  return cd, err
+				}, timeout, interval).ShouldNot(BeNil())
+
+				By("Getting ServiceAccount")
+				sa := &corev1.ServiceAccount{}
+				Eventually(func() (*corev1.ServiceAccount, error) {
+				  err := k8sClient.Get(context.Background(),
+				  	testsutil.GetResourceKey(config.CSIControllerServiceAccount, found.Name, found.Namespace), sa)
+				  return sa, err
+				}, timeout, interval).ShouldNot(BeNil())
+
+				By("Getting ClusterRole")
+				cr := &rbacv1.ClusterRole{}
+				for _, clusterRole := range clusterRoles {
+					Eventually(func() (*rbacv1.ClusterRole, error) {
+						err := k8sClient.Get(context.Background(),
+							testsutil.GetResourceKey(clusterRole, found.Name, ""), cr)
+						return cr, err
+					}, timeout, interval).ShouldNot(BeNil())
+				}
+
+				By("Getting ClusterRoleBinding")
+				crb := &rbacv1.ClusterRoleBinding{}
+				for _, clusterRoleBinding := range clusterRoleBindings {
+					Eventually(func() (*rbacv1.ClusterRoleBinding, error) {
+						err := k8sClient.Get(context.Background(),
+							testsutil.GetResourceKey(clusterRoleBinding, found.Name, ""), crb)
+						return crb, err
+					  }, timeout, interval).ShouldNot(BeNil())
+				}
+
+				By("Getting controller StatefulSet")
+				controller := &appsv1.StatefulSet{}
+				Eventually(func() (*appsv1.StatefulSet, error) {
+				  err := k8sClient.Get(context.Background(),
+				  	testsutil.GetResourceKey(config.CSIController, found.Name, found.Namespace), controller)
+				  return controller, err
+				}, timeout, interval).ShouldNot(BeNil())
+				checkContainersImages(controller.Spec.Template.Spec)
+
+				By("Getting node DaemonSet")
+				node := &appsv1.DaemonSet{}
+				Eventually(func() (*appsv1.DaemonSet, error) {
+				  err := k8sClient.Get(context.Background(),
+				  	testsutil.GetResourceKey(config.CSINode, found.Name, found.Namespace), node)
+				  return node, err
+				}, timeout, interval).ShouldNot(BeNil())
+				checkContainersImages(node.Spec.Template.Spec)
+
+				close(done)
+			  }, timeout.Seconds())
+		})
+	})
+})
+
+func checkContainersImages(podSpec corev1.PodSpec) {
+	Expect(len(podSpec.Containers)).To(BeNumerically(">", 0))
+	for _, container := range podSpec.Containers {
+		image, ok := containersImages[container.Name]
```

Comment:
what if a container is missing in the pod? shouldn't we ensure all containers in the cr exist in any pod?

Reply:
I think it's enough for each cr image to be in ANY pod, so no need to maintain said list

Reply:
i.e. only assert:
set(<cr-images>) ⊂ set(<running-containers>)
(if not, print all not-running cr-images)

Reply:
done

Reply:
I meant that we don't need other image/container assertions.
e.g. no need to fail when there is a deployed container that isn't in the cr yaml (we are not required to make every deployed container image configurable)

Reply:
First of all, I think we should support that every image that we deploy will be configurable because our product can be used in an offline env.
Second of all, so I don't understand what you want me to check if we don't need image/container assertions.

Reply:
1. good point
2. I suggested to (only) assert [this](https://github.com/IBM/ibm-block-csi-operator/pull/248#discussion_r835891220)
and if I understand correctly, this code checks the opposite "direction" (set(`<running-containers>`) ⊂ set(`<cr-images>`)).

but I agree both "directions" are important, so could we just assert set(`<running-containers>`) == set(`<cr-images>`) (and if not, print the (symmetric) difference)

Reply:
done